Dashboard: enhance SS locations and usage pages

jraddaoui opened the pull request:
Dashboard: unify SS locations administration pages

    - Remove transfer source locations URL, link and view.
    - Rename 'AIP storage locations' link to 'Storage locations'.
    - Change 'Storage locations' page title and add page description.
    - Show all related SS locations for this pipeline in this page, except
      currently processing, AIP recovery and SS internal purposes and
      disabled locations.
    - Sort locations by purpose and show it in human readable form.
    - Show formated used and quota values for AIP and DIP storage locations.
    - Add tests.

Dashboard: improve processing storage usage page

    To avoid timeouts on the first load of this page, the usage data is only
    calculated when the `calculate` get parameter is evaluated as `True`.
    When the usage data is not calculated, the page shows a description and
    a button to reload the page calculating the usage. When the usage is
    calculated, the page will display a general information section and a
    table of clearable directories from within the shared path.

    - Show the root of the filesystem where the shared directory is located.
    - Only show the size of the root path.
    - Rename `_usage_dirs`function to`_get_shared_dirs` and change default
     `calculate_usage` to`False`.
    - Simplify directory list generation and only include clearable dirs.
    - Add`--one-file-system` option to `du` calls.
    - Fix translations.
    - Cosmetic changes.
    - Add tests.

Connects to https://github.com/archivematica/Issues/issues/312.

jraddaoui commented:
Hi @ross-spencer, this is ready for an initial review.

I've sent an email to the team with some notes about this changes and the missing tests (related to https://github.com/archivematica/Issues/issues/783).

ross-spencer commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -29,6 +29,7 @@
 from django.http import Http404, HttpResponseNotAllowed, HttpResponseRedirect
 from django.shortcuts import redirect, render
 from django.template import RequestContext
+from django.template.defaultfilters import filesizeformat
```
This is cool! Useful to know about!

ross-spencer commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -176,113 +177,191 @@ def _atom_levels_of_description_sort_adjust(level_id, sortorder="promote"):
         messages.warning(
             request,
             _(
-                "Error retrieving locations: is the storage server running? Please contact an administrator."
```
I can't see a difference between left and right here so I think this would have been left as-is by `black` given this is a `qa/1.x` branch, so I don't think it would have been a problem to leave as-is. No need to change it now though.

jraddaoui replied:
Yes, black doesn't split that string to fit the usual max line length on its formatting, so I did it manually to make it more readable, I don't really know why I split it in three, it's in two lines now.

ross-spencer commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -176,113 +177,191 @@ def _atom_levels_of_description_sort_adjust(level_id, sortorder="promote"):
-    system_directory_description = "Available storage"
-    return render(request, "administration/locations.html", locals())
+    # Sort by purpose
+    locations.sort(key=lambda loc: loc["purpose"])
```
Nice touch sorting them!

ross-spencer commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -176,113 +177,191 @@ def _atom_levels_of_description_sort_adjust(level_id, sortorder="promote"):
-    Return page summarizing storage usage
-    """
-    usage_dirs = _usage_dirs()
+    calculate_usage = request.GET.get("calculate", False)
```
I think technically this evaluates a string, so as long as the string isn't `""` e.g. `http://127.0.0.1:62080/administration/usage/?calculate=do_not_calculate` you'll generate the report. Not a big deal, but you could consider making it more strict or update the doc string. 

Confirmed when modifying in the unit test as well, so if it's important we might consider how to make those more robust to reflect the query string we want to allow.

jraddaoui replied:
Great point @ross-spencer, I'll improve this.

ross-spencer commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -176,113 +177,191 @@ def _atom_levels_of_description_sort_adjust(level_id, sortorder="promote"):
+                    "description": "Transfers",
+                    "path": os.path.join(shared_path, "completed", "transfers"),
+                },
+            ),
```
More a usage question... what processing configuration setting do we need to be able to measure this? i.e. when is `/var/archivematica/sharedDirectory/completed/transfers` populated for any length of time? (I've tried not sending a transfer to backlog so it sits at the decision point before that, but that doesn't seem right).

jraddaoui replied:
I really don't know. We chatted about that in the last meeting and if I remember correctly this was going to be tested (not sure if in the tests for this feature or in another test instance). cc: @sallain.

sallain replied:
I also don't have an answer for this, unfortunately.

ross-spencer commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -176,113 +177,191 @@ def _atom_levels_of_description_sort_adjust(level_id, sortorder="promote"):
-            dir_spec["used"] = _usage_get_directory_used_bytes(space_path)
+    if calculate_usage:
+        for id, dir_spec in dirs.items():
+            dir_spec["used"] = _usage_get_directory_used_bytes(dir_spec["path"])
```
Much more elegant! Lovely.

ross-spencer commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -176,113 +177,191 @@ def _atom_levels_of_description_sort_adjust(level_id, sortorder="promote"):
         # Split value by whitespace and size (in blocks)
         size = usage_summary.split()[1]
-
         return int(size) * 1024
```
Would it be appropriate to use `filesizeformat` here too?

jraddaoui replied:
I think it's better to return bytes in here and do the formatting in the view/template. It'll make this function more re-usable.

ross-spencer commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -310,91 +390,53 @@ def _usage_get_directory_used_bytes(path):
+    if not dir_info:
+        raise Http404
+
+    prompt = _("Clear %(dir)s?") % {"dir": dir_info["description"]}
```
I'm still getting used to this syntax, it's a little more involved than I think we normally use in the AM code-base. If it were a plain-string, it might look like: `"Clear {}?".format(dir_info["description"])`, If i remember correctly the `_(...)` enables translation?

jraddaoui replied:
Yes, `_()` is usually [the translations wrapper](https://github.com/artefactual/archivematica/blob/stable/1.9.x/src/dashboard/src/components/administration/views.py#L33). By using this syntax it gives a little more context to the translator about the string and allows to change the placeholder location easily. This may not be the best example, but look at [this one from the docs](https://docs.djangoproject.com/en/1.8/topics/i18n/translation/#standard-translation) (at the bottom of the "Standard translation" section).

ross-spencer commented on src/dashboard/src/templates/administration/locations.html:
```diff
@@ -16,22 +16,34 @@
-      <h3>{{ system_directory_description }}</h3>
+      <h3>{% trans "Available storage locations" %}</h3>
+
+      <p>{% trans "Storage locations are configured by an administrator in the Storage Service, Archivematica's back-end storage manager." %}</p>
```
I wonder if this can be rephrased, or if I am reading this correctly? Perhaps we can ask Sara? In `"Storage locations are configured by an administrator in the Storage Service, Archivematica's back-end storage manager."` I feel like the subject is the 'administrator' and we're describing them as "Archivematica's back-end storage manager" :sweat_smile: 

I might suggest:

> "Storage locations are configured by an administrative user in the Archivematica Storage Service."

jraddaoui replied:
I'm really open to any suggestion on any of the sentences of this page ;)

cc: @sallain ^^

sromkey replied:
I agree with Ross's suggestion, I think that's fine!

sallain replied:
@jraddaoui I think that Ross' suggestion is a little clearer! Tbh it's not a big deal though, the sentence reads fine without the change.

ross-spencer commented on src/dashboard/src/templates/administration/usage.html:
```diff
@@ -18,32 +18,97 @@
+          Archivematica pipeline configuration and in the Storage Service, Archivematica's
+          back-end storage manager. The shared directory configuration path must match the
+          currently processing location path used for this pipeline in the Storage Service
+          for this page to work properly.
```
Reading this, I'm not quite sure what to do as a user/administrator for `The shared directory configuration path must match the currently processing location path`. I wonder if there could be an example on this page too? 

That being said, I wonder if there is a terminology issue? As in, if I look at the storage service, I haven't actually got a field I can fill in called 'shared directory'; either in the locations or the storage space forms? I might be missing it though...

jraddaoui replied:
When a pipeline is registered in the SS, it's shared directory is passed to create the currently processing and other locations inside it by default. However, that can be changed later in the SS modifying the locations path. In the case where the pipeline's shared directory and the SS CP location path don't match, I think there will be a lot more issues in the pipeline than just this page.

However, I think this paragraph was in here to give a little more sense to the relation in between the processing storage usage and the shared path.

cc: @sallain

sromkey replied:
Since it sounds like the whole pipeline would be borked if the paths didn't match maybe it's more clear to just remove the sentence.

ross-spencer commented on src/dashboard/tests/test_administration_storage.py:
```diff
@@ -0,0 +1,150 @@
+        # One of them has an unlimited quota set
+        self.assertEqual(
+            len([loc for loc in locations if loc["quota"] == "unlimited"]), 1
+        )
```
This is an aside... (it works well in the UI!): what does the quota setting actually do in the storage service? Does it trigger any actions?

jraddaoui replied:
It's being checked/updated when packages are moved from/to a SS location and it should cause an error if it's reached in that process, but I think it still needs some enhancements. There are some minor comments about it in https://github.com/archivematica/Issues/issues/406.

ross-spencer reviewed (changes requested):
@jraddaoui this looks really great. There are no significant code changes required from my perspective. I've left some comments to ask for clarification (one or two for personal learning!), I think this is pretty much good to go after that.

jraddaoui commented on src/dashboard/tests/test_administration_usage.py:
```diff
@@ -17,10 +17,16 @@ def setUp(self):
-        self.assertFalse(response.context["calculate_usage"])
-        self.assertIn('<a href="?calculate=true"', response.content)
-        self.assertIn("Calculate disk usage", response.content)
+        for calculate in [None, "False", "false", "no", "0", "random"]:
```
Parametrization doesn't work in `unittest.TestCase` sub-classes: https://docs.pytest.org/en/latest/unittest.html#pytest-features-in-unittest-testcase-subclasses :(

ross-spencer replied:
Oh yeah, this is a thing :slightly_frowning_face:  another small thing we can look at in future though :slightly_smiling_face:

jraddaoui commented:
Back to you @ross-spencer and @sallain.

I've moved the tests to the administration app in the last commit, based on @sevein's feedback from https://github.com/artefactual/archivematica/pull/1448.

ross-spencer commented:
Sounds great, thanks Radda.

I feel like I should wait for Sara to have a once over of the messages she
is CCd in and then I can run through this for a brief final review. If that
works?

On Thursday, July 11, 2019, José Raddaoui Marín <notifications@github.com>
wrote:

> Back to you @ross-spencer <https://github.com/ross-spencer> and @sallain
> <https://github.com/sallain>.
>
> I've moved the tests to the administration app in the last commit, based
> on @sevein <https://github.com/sevein>'s feedback from #1448
> <https://github.com/artefactual/archivematica/pull/1448>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/artefactual/archivematica/pull/1314?email_source=notifications&email_token=AAOLCXA3UWDR7FZB4OANZRLP64ZWXA5CNFSM4GKBBDEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZWXFHI#issuecomment-510489245>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOLCXDJODDPH5RFUASIVGLP64ZWXANCNFSM4GKBBDEA>
> .
>


-- 
Ross Spencer
Software Developer, Artefactual Systems Inc.
www.artefactual.com

ross-spencer commented on src/dashboard/src/components/administration/tests/test_storage.py:
```diff
@@ -8,8 +8,8 @@
-class TestAdministrationStorage(TestCase):
-    fixtures = [os.path.join(THIS_DIR, "fixtures", "test_user.json")]
+class TestStorage(TestCase):
+    fixtures = [os.path.join(THIS_DIR, "../../../../tests/fixtures", "test_user.json")]
```
I don't know the prettiest way to do this (or a better way!), but I feel like we're more idiomatically using the following approach in other parts of the system:
```python
    dashboard_path = os.path.dirname(
        os.path.dirname(
            os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
        )
    )
    fixtures_path = os.path.join(
        THIS_DIR, dashboard_path, os.path.join("tests", "fixtures"), "test_user.json"
    )
    fixtures = [fixtures_path]
```
This should just work, but if we don't elect to change this now, we can always do it in future. Same for `test_usage.py` below.

jraddaoui replied:
Hi @ross-spencer, I recently fixed the fixtures situation in https://github.com/artefactual/archivematica/commit/c63c1195e61b39cc53ad9f176059b3dae854b04e, I just didn't rebase this work after those changes. Will do and thanks for the remainder ;)

ross-spencer replied:
Ah, I had a really loose memory of that @jraddaoui but couldn't get it back. Feels like such a long time ago! Anyway, looking forward to pushing on :slightly_smiling_face:

ross-spencer commented on src/dashboard/src/components/administration/tests/test_usage.py:
```diff
@@ -0,0 +1,53 @@
+        self.client.login(username="test", password="test")
+        helpers.set_setting("dashboard_uuid", "test-uuid")
+
+    def test_no_calculation(self):
```
Nice test, the corresponding change in the dashboard works nicely with a few different values too :+1:

ross-spencer commented on src/dashboard/src/components/administration/tests/test_usage.py:
```diff
@@ -0,0 +1,53 @@
+                response = self.client.get(
+                    "/administration/usage/?calculate=%s" % calculate
+                )
+            self.assertFalse(response.context["calculate_usage"])
```
This is cool @jraddaoui. The context object is a `django.test.utils.ContextList` I had to look up the [docs](https://docs.djangoproject.com/en/2.2/topics/testing/tools/#django.test.Response.context) but it looks really useful.

ross-spencer reviewed (approved):
Thanks @jraddaoui. Thanks for making the change to the string matching. Thanks for fielding my questions too. Some nice things I'm learning about reading through this. This looks good and is working in my deployment. 

I've left one comment for loading fixtures, but everything's working so up to you. 

Normally we (I?) don't do this,but I've resolved the conversations too so it's just the terminology/language ones left for @sallain. And there's one that might form part of a test-plan for testing this location: `/var/archivematica/sharedDirectory/completed/transfers`, i.e. figuring out what data populates it! :slightly_smiling_face: 

Equally, and all being well, these are things that will crop up in UAT, and so when we get the go-ahead for `1.11` this seems good to rebase and merge. Nice work.

jraddaoui committed:
Dashboard: enhance SS locations and usage pages

Unify SS locations administration pages:

- Remove transfer source locations URL, link and view.
- Rename 'AIP storage locations' link to 'Storage locations'.
- Change 'Storage locations' page title and add page description.
- Show all related SS locations for this pipeline in this page, except
  currently processing, AIP recovery and SS internal purposes and
  disabled locations.
- Sort locations by purpose and show it in human readable form.
- Show formated used and quota values for AIP and DIP storage locations.
- Add tests.

Improve processing storage usage page:

To avoid timeouts on the first load of this page, the usage data is only
calculated when the `calculate` get parameter is evaluated as `True`.
When the usage data is not calculated, the page shows a description and
a button to reload the page calculating the usage. When the usage is
calculated, the page will display a general information section and a
table of clearable directories from within the shared path.

- Show the root of the filesystem where the shared directory is located.
- Only show the size of the root path.
- Rename `_usage_dirs`function to`_get_shared_dirs` and change default
  `calculate_usage` to`False`.
- Simplify directory list generation and only include clearable dirs.
- Add`--one-file-system` option to `du` calls.
- Fix translations.
- Cosmetic changes.
- Add tests.

Squashed commit of the following:

commit a65781150fe60b4adfeceed2e7b31519cd7ee949
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Oct 18 11:38:45 2019 +0200

    Fix wording

commit cdecde614b5fe40c61480ecaee69f6b685a359de
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Thu Sep 5 11:00:51 2019 +0200

    Improve fixtures declaration after rebase

commit e5dcaafe7d0202c546e96e80e4ec2388bc57cd61
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Thu Jul 11 15:35:48 2019 +0200

    Move tests to package within administration app

commit e58d203f536a692bc2d345bc02bde2bc974b1b1f
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Thu Jul 11 15:24:25 2019 +0200

    Improve calculate boolean validation

commit 6ce24f0b0459a381b17b157cd5ad2f0ed6654a30
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Thu Jul 11 14:50:38 2019 +0200

    Re-format string

commit af53844ccc2a54ea668ad5489a1bfdfd4ef4f7fb
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon Jul 8 19:24:03 2019 +0200

    Dashboard: improve processing storage usage page

    To avoid timeouts on the first load of this page, the usage data is only
    calculated when the `calculate` get parameter is evaluated as `True`.
    When the usage data is not calculated, the page shows a description and
    a button to reload the page calculating the usage. When the usage is
    calculated, the page will display a general information section and a
    table of clearable directories from within the shared path.

    - Show the root of the filesystem where the shared directory is located.
    - Only show the size of the root path.
    - Rename `_usage_dirs`function to`_get_shared_dirs` and change default
      `calculate_usage` to`False`.
    - Simplify directory list generation and only include clearable dirs.
    - Add`--one-file-system` option to `du` calls.
    - Fix translations.
    - Cosmetic changes.
    - Add tests.

commit ed316c03728037c145d3b147972d24d4fae9190a
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon Jul 8 19:23:45 2019 +0200

    Dashboard: unify SS locations administration pages

    - Remove transfer source locations URL, link and view.
    - Rename 'AIP storage locations' link to 'Storage locations'.
    - Change 'Storage locations' page title and add page description.
    - Show all related SS locations for this pipeline in this page, except
      currently processing, AIP recovery and SS internal purposes and
      disabled locations.
    - Sort locations by purpose and show it in human readable form.
    - Show formated used and quota values for AIP and DIP storage locations.
    - Add tests.
